### PR TITLE
[Serve][Doc] Triton server integration

### PR DESCRIPTION
**Pull request by sihanwang41**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+(serve-triton-server-integration)=
+
+# Serving a Resenet model with Triton Server
+This guide shows how to serve models with [NVIDIA Triton Server](https://github.com/triton-inference-server/server) using Ray Serve.
+
+## Installation
+Here is the Dockerfile example for installing Triton Server with Ray Serve.
```
Comment (sihanwang41) [inline doc/source/serve/tutorials/triton-server-integration.md:7]: wait for dockerfile from nvidia.

**Comment by edoakes** [comment]
Ping me when ready for review

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from fastapi import FastAPI
+from ray import serve
+from tritonserver_api import TritonServer
+
+
+# 1: Define a FastAPI app and wrap it in a deployment with a route handler.
+app = FastAPI()
+
+
+@serve.deployment(route_prefix="/", ray_actor_options={"num_gpus": 1})
```
Comment (akshay-anyscale) [inline doc/source/serve/tutorials/triton-server-integration.md:34]: can we add some autoscaling options?

Comment (sihanwang41) [reply doc/source/serve/tutorials/triton-server-integration.md]: added, just finish testing, it is working normally.

**Comment by sihanwang41** [comment]
> Ping me when ready for review

Sounds good, let me know if you think anything needed to be added. The major thing waiting from nvidia is Dockerfile part.

**Comment by sihanwang41** [comment]
@akshay-anyscale PTAL

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    ├── 1
+    │   └── model.plan
+    └── config.pbtxt
+```
+
+The model repository contains three models: `stable_diffusion`, `text_encoder` and `vae`. Each model has a `config.pbtxt` file and a model file. The `config.pbtxt` file contains the model configuration, which is used to describe the model type and input/output formats. (You can learn more about model config file [here](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_configuration.md)). To get config files for our example, you can download them from [here](https://github.com/triton-inference-server/tutorials/tree/main/Conceptual_Guide/Part_6-building_complex_pipelines/model_repository). We use `1` as the version of each model. The model files are saved in the version directory.
+
+
+## Start Ray Serve with the Triton Server
+Triton Server provides python API to start the Triton Server instance. You can use the `nvcr.io/nvidia/tritonserver:23.12-py3` image which already have the Triton Server python API library installed.
```
Comment (akshay-anyscale) [inline doc/source/serve/tutorials/triton-server-integration.md:92]: Can we add the complete instructions for how to build an image that has both triton and ray?

Comment (sihanwang41) [reply doc/source/serve/tutorials/triton-server-integration.md]: done.

**Comment by sihanwang41** [comment]
Hi @edoakes, the pr is updated, PTAL.

**Review by akshay-anyscale — APPROVED** [review]

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
+(serve-triton-server-integration)=
+
+# Serving model with Triton Server in Ray Serve
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:3]:
```suggestion
# Serving models with Triton Server in Ray Serve
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+This guide shows how to build an application with stable diffusion model using [NVIDIA Triton Server](https://github.com/triton-inference-server/server) in Ray Serve.
+
+## Preparation
+
+### Installation
+It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image which already have the Triton Server python API library installed, and install the ray serve lib by `pip install "ray[serve]"` inside the image.
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:9]:
```suggestion
It is recommended to use the `nvcr.io/nvidia/tritonserver:23.12-py3` image which already has the Triton Server python API library installed, and install the ray serve lib by `pip install "ray[serve]"` inside the image.
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+### Build Model
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:11]:
```suggestion
### Build and export a model
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+For this application, the encoder are exported to ONNX format and the stable diffusion model are exported to be TensorRT engine format which is being compatible with Triton Server.
+Here is the example to exporting models to be ONNX format.([source](https://github.com/triton-inference-server/tutorials/blob/main/Triton_Inference_Server_Python_API/scripts/stable_diffusion/export.py))
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:13]:
```suggestion
For this application, the encoder is exported to ONNX format and the stable diffusion model is exported to be TensorRT engine format that is compatible with Triton Server.
Here is the example to export models to in ONNX format.([source](https://github.com/triton-inference-server/tutorials/blob/main/Triton_Inference_Server_Python_API/scripts/stable_diffusion/export.py))
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        "input_ids": {0: "batch", 1: "sequence"},
+    },
+    opset_version=14,
+    do_constant_folding=True,
+)
+```
+
+From the script, the outputs are `vae.onnx` and `encoder.onnx`.
+
+After the ONNX model exported, converting the ONNX model to the TensorRT engine serialized file. ([Details](https://github.com/NVIDIA/TensorRT/blob/release/9.2/samples/trtexec/README.md?plain=1#L22) about trtexec cli)
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:66]:
```suggestion
After the ONNX model exported, convert the ONNX model to the TensorRT engine serialized file. ([Details](https://github.com/NVIDIA/TensorRT/blob/release/9.2/samples/trtexec/README.md?plain=1#L22) about trtexec cli)
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+```bash
+trtexec --onnx=vae.onnx --saveEngine=vae.plan --minShapes=latent_sample:1x4x64x64 --optShapes=latent_sample:4x4x64x64 --maxShapes=latent_sample:8x4x64x64 --fp16
+```
+
+### Prepare Model Repository
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:71]:
```suggestion
### Prepare the model repository
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+│   ├── 1
+│   │   └── model.onnx
+│   └── config.pbtxt
+└── vae
+    ├── 1
+    │   └── model.plan
+    └── config.pbtxt
+```
+
+The model repository contains three models: `stable_diffusion`, `text_encoder` and `vae`. Each model has a `config.pbtxt` file and a model file. The `config.pbtxt` file contains the model configuration, which is used to describe the model type and input/output formats. (You can learn more about model config file [here](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_configuration.md)). To get config files for our example, you can download them from [here](https://github.com/triton-inference-server/tutorials/tree/main/Conceptual_Guide/Part_6-building_complex_pipelines/model_repository). We use `1` as the version of each model. The model files are saved in the version directory.
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:91]:
```suggestion
The model repository contains three models: `stable_diffusion`, `text_encoder` and `vae`. Each model has a `config.pbtxt` file and a model file. The `config.pbtxt` file contains the model configuration, which is used to describe the model type and input/output formats (you can learn more about model config file [here](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_configuration.md)). To get config files for our example, you can download them from [here](https://github.com/triton-inference-server/tutorials/tree/main/Conceptual_Guide/Part_6-building_complex_pipelines/model_repository). We use `1` as the version of each model. The model files are saved in the version directory.
```

File: doc/source/serve/tutorials/triton-server-integration.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+
+## Start Ray Serve with the Triton Server
```
Comment (edoakes) [inline doc/source/serve/tutorials/triton-server-integration.md:94]:
```suggestion
## Start the Triton Serve inside of a Ray Serve application
```

**Review by edoakes — COMMENTED** [review]
just nits

**Review by edoakes — APPROVED** [review]

**Pull request merged**